subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/DefaultScriptCompilationHandler.java

```diff
@@ -380,4 +388,41 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
             return super.loadClass(name, resolve);
         }
     }
+
+    private static class InstrumentingScriptClassLoader extends ScriptClassLoader implements InstrumentingClassLoader {
+        private final ThreadLocal<Throwable> pendingException = new ThreadLocal<>();
```

Comment:
🎨 Perhaps a `InstrumetingClassLoaderMixin` implementation would be useful so this class could mostly delegate to the mixin implementation?